### DISCUSSION TITLE
Flux2 GUI Proposal

### DISCUSSION THREAD

**Discussion by jpellizzari** [fluxcd/flux2#503] [Proposals] [answered]
# Flux GUI Proposal

This proposal seeks to establish the ground-work for beginning UX and architecture design for a Flux graphical user interface. This proposal assumes that a "UI" in this context is running a web browser or similar environment.

## Guiding Principles

These are the main factors and constraints to consider when designing and implementing the Flux UI

- Kubernetes objects are the API and "persistence" layers for the UI
- All write operations must done via `git` (reconcilliation is not consiered a write operation)
- The UI should always seek to push users towards GitOps best-practices
- We should **not** seek to replace established tools and best-practices for write operations where applicable (ie no code editors in the UI)
- The UI should be theme-able
- The UI must observe RBAC and Kubernetes client best-practices, including performance considerations

### MVP user flows

- Browse flux-related objects and understand their status (read-only)
- Reconcile a Source
- Reconcile a Kustomization
- Reconcile a Helm Release

### Near-term aspirational user flows

- Change the container image of a Deployment
- Suspend and resume reconcillation
- Add/edit/remove a notification reciever

### Long-term aspirational user flows

- Navigate from a Deployment to a corresponding CI provider build number that last modified or interacted with the Deployment

## Entry points

The help guide design and implementation, it is helpful to describe how users will enter into the Flux UI experience. Describing the aspirational state of the UI helps inform the design and implementation.

### MVP entry points

For the initial launch of the UI, the only way to enter into the experience will be to
run the UI locally and inherit the users `.kubeconfig`:

```shell
$ flux run ui
Flux UI is now available on http://localhost:4567
```

**Why?**: This represents the simplest possible implementation and allows us to get the initial code-paths in place before tackling harder authn situations.

### Aspirational entry points

- Install the UI on the cluster permanently (via `git` operations):

  1. `$ flux install ui`
  2. `$ kubectl port-forward fluxui 8080:8080`

  **Why?**: We want the UI to be a permanent fixture for day-to-day operations for teams that adopt Flux. We also want to lower the barriers-to-entry for users that are not intimately familiar with GitOps practices.

* Install the UI as a `react` component in another UI project:

  1.  `$ npm i @fluxcd/fluxui`
  2.  `import { FluxComponent } from ‘@fluxcd/fluxui’`
  3.  `<MyApp><FluxComponent /></MyApp>`

  **Why?**: As a means to increase adoption, we want this Flux UI to be compatible with other projects in the ecosystem.

**Comment by scottrigby** [comment]
Nice 👏  Some initial thoughts on a few of the points above:

### Objects

> Browse flux-related objects and understand their status

I love this, and have also been imagining a UI could mirror the concepts implemented with v2 architecture, at least generally insofar as it helps end users with continuity between their configuration and what they should expect to see eventually reconciled.

Somewhere I remember @cdavisafc describing "Source Controllers" and "Delivery Controllers" (Kustomize controller, Helm controller, etc). It may not be necessary to see controller info, but I wonder if it's valuable to consider categorizing the Flux CRD objects for end users in a similar way? Something like:

```yaml
Sources:
- GitRepository
- HelmRepository
- Bucket

Delivery:
- Kustomization
- HelmRelease
- CAPI (or whatever else)

Notifications:
- events
- providers 
- alerts
- receivers
```

### Fronend components

> Install the UI as a react component in another UI project
> Why?: As a means to increase adoption, we want this Flux UI to be compatible with other projects in the ecosystem.

From the chat today this was my big takeaway. @stealthybox described the component idea and noted a potential challenge that native UIs may want interconnected and deep linking. Perhaps instead of a single component, there could be multiple, reusable components (with configurable links etc) to play nicely with various existing UI frameworks? Somehow componentized, pluggable, and extensible?

Spitballing from the meeting and your proposal above, maybe the UI repo could hold:
1. React components;
2. A thin wrapper that allows users to spin up those components into a basic, functional UI they can launch from the command line; and
3. Other implementations could use the same tools to build their own wrappers and use the same shared components for native UIs?

### Auth

One other fun note. I've participated a bit in the CNCF Artifact Hub project, and I really like how their UI allows users to manage access with sane defaults, but also optional custom OPA policies. IDK if we ever want to go here, but depending this may be worth discussing for the future.

Looking forward to seeing how this discussion unfolds 👀 🍿 😄

**Reply by jpellizzari** [reply]
> Perhaps instead of a single component,

I would be okay with saying we provide "one or many" components.

> A thin wrapper that allows users to spin up those components into a basic, functional UI they can launch from the command line; and

My thoughts exactly.

**Reply by jpellizzari** [reply]
```
Delivery:
- Kustomization
- HelmRelease
- CAPI (or whatever else)
```

I would suggest we wait for the larger project to adopt this mental model before pushing it forward in the UI

**Reply by phillebaba** [reply]
I think the biggest challenge is going to be how we authenticate users. Having to implement access control separate to RBAC is kind of a pain in the ass and is the main reason I have avoided GUIs for so long. 

It might be good to look at how backstage is trying to solve the issue of authenticating their users to different Kubernetes clusters.
https://github.com/backstage/backstage/issues/2461

**Reply by thmzlt** [reply]
Regarding authentication, I like that in Argo and Grafana you can set it up for a GitHub organization/team via OIDC and avoid keeping a database of users.

**Reply by shanesveller** [reply]
Maintaining a local database of users is less the challenge than reimplementing most of the same features and scope of the RBAC system Kubernetes already has - a write-capable UI will need RBAC controls as not all users will be permitted to make all actions in all namespaces. If it can't reuse Kube identity and RBAC, that's a lot of redundant work to reach feature parity. A satisfactory auth story is crucial for potential adoption or such a tool will be relegated to dev environments only if it's not handled with grace.

**Reply by jpellizzari** [reply]
The subject of auth will probably get its own discussion thread at some point. The initial implementation will be in line with similar dev tools, such as Octant and Lens.

> A satisfactory auth story is crucial for potential adoption or such a tool will be relegated to dev environments only if it's not handled with grace.

I think this^^ is a very succinct way to sum it up. We want this UI to be a fixture in users' production environments.

**Comment by jpellizzari** [comment]
I would also like to enumerate any performance considerations we should think about, specifically around not overwhelming the Kubernetes API server.

Maybe the folks familiar with building/tuning controllers can weigh in? Do we need to think about caching now, or can that be deferred?

We eventually want this UI to be usable by dozens, then hundreds of users simultaneously. What can we do in the initial MVP to achieve that vision?

**Reply by squaremo** [reply]
For the MVP I would simply be modest (and honest) with the scale that is supported, and establish the expectation that the view will never reflect the cluster state perfectly. Then you have a lot of leeway to find where the bottlenecks actually are, with the architecture and features as designed.

Having said that, I think it would be a decent punt for a _local_ GUI to use watches and keep a local cache -- the GUI will be doing a lot of listing, I imagine. The controller-runtime client may help here, if you are concerned only with a fixed set of types (e.g., all the GOTK API types). I think this can go a fairly long way, but if you are nervous about running a large number of watches, do an experiment to see how it scales.

**Comment by jpellizzari** [comment]
A quick architecture diagram I put together (v1):

![fluxui](https://user-images.githubusercontent.com/2802257/100641006-d2300d80-32eb-11eb-9268-c0c03f389cb7.png)

**Comment by jpellizzari** [comment]
Here are some basic wireframes for the proposed layout for the MVP version of the GUI:

![New Wireframe 1 (1)](https://user-images.githubusercontent.com/2802257/100762899-5d230d80-33a9-11eb-8a99-2b60ee39aa1b.png)
![New Wireframe 1 copy](https://user-images.githubusercontent.com/2802257/100762909-60b69480-33a9-11eb-9e9e-f15d6c31223f.png)

**Comment by jpellizzari** [comment, marked as answer]
I would also like to propose that we move forward with creating a git repo under the FluxCD org. I don't have strong feelings about the name. Some options: `fluxcd/ui`, `fluxcd/fluxui`, `fluxcd/gui`

**Reply by squaremo** [reply]
Let's make this happen @fluxcd/oversight-committee 

✅  for creating the repo from me.

`fluxcd/gui`, `fluxcd/webui` are my preferred names.

**Reply by hiddeco** [reply]
:heavy_check_mark: from me too, and a :heavy_plus_sign: for `fluxcd/webui`.

**Reply by stefanprodan** [reply]
✔️  `fluxcd/webui`

**Reply by squaremo** [reply]
@jpellizzari Happy with `fluxcd/webui`?

**Reply by jpellizzari** [reply]
@squaremo yep!

**Comment by jpellizzari** [comment]
I would also propose that aesthetically, the GUI borrows heavily from the [fluxcd website](https://fluxcd.io/) (colors, typography, etc)

**Comment by cdavisafc** [comment]
@jpellizzari and all - nice job. 

> I remember @cdavisafc describing "Source Controllers" and "Delivery Controllers"

I actually talk about delivery controllers and runtime controllers. Ultimately GitOps, and hence Flux, will build out not only the delivery capabilities but enhanced runtime capabilities - note that Flagger is listed in the "runtime controllers" bucket.

![Screen Shot 2020-12-01 at 5 34 26 AM](https://user-images.githubusercontent.com/1652270/100822086-7efead80-3406-11eb-8f02-214060e2b80b.png)

**Reply by scottrigby** [reply]
Thanks @cdavisafc this is what I was remembering 👍

**Comment by joebowbeer** [comment]
I suggest looking at octant and other k8s dashboards for inspiration and optimization techniques.

In fact, the Flux UI could be implemented as an octant plugin.
